lib/truffle/truffle/patching.rb

```diff
@@ -1,34 +1,119 @@
+              "bundler/current_ruby"                    => :after,
+              "bundler/dependency"                      => :after,
+              "bundler/fetcher/downloader"              => :after,
+              "bundler/ruby_version"                    => :after,
+              "bundler/source/rubygems"                 => :after,
+
+              "rubygems"                                => :before,
+              "rubygems/ext"                            => :after,
+              "rubygems/package"                        => :after,
+              "rubygems/remote_fetcher"                 => :after,
+              "rubygems/request"                        => :after,
+              "rubygems/request_set/gem_dependency_api" => :after,
+
+              "rspec/support"                           => :after,
+
+              "unf/normalizer_cruby"                    => :before,
+  }
+
+  PATCHES.each do |file, _|
+    unless File.exist? File.join(DIR, file) + '.rb'
+      raise "file #{file} is missing in #{DIR}"
+    end
+  end
+
+  Dir["#{DIR}/**/*.rb"].each do |file|
+    relative_path = file[(DIR.size + 1)..-4]
+    unless PATCHES.key? relative_path
+      raise "file #{relative_path} is not declared in PATCHES"
+    end
+  end
+
+  def self.find_absolute_path(relative_path)
+    $LOAD_PATH.each do |base|
+      candidate = File.join(base, relative_path) + '.rb'
+      return candidate if File.exist? candidate
+    end
+    raise "absolute path for #{relative_path} was not find in #{$LOAD_PATH}"
+  end
+
+  def self.find_relative_path(absolute_path)
+    absolute_path_with_suffix = if absolute_path.end_with?('.rb')
+                                  absolute_path
+                                else
+                                  absolute_path + '.rb'
+                                end
+
+    raise "file #{absolute_path} does not exist" unless File.exist? absolute_path_with_suffix
+    $LOAD_PATH.each do |base|
+      if absolute_path_with_suffix.start_with?(base)
+        return File.basename absolute_path_with_suffix[(base.size)..-4]
+      end
+    end
+
+    return nil
+  end
+
+  def self.required?(relative_path)
+    $LOADED_FEATURES.include? find_absolute_path relative_path
+  end
 end
 
 module Kernel
 
   private
 
-  alias_method :require_without_truffle_patching, :require
+  alias_method :require_without_truffle_patching, :gem_original_require
 
-  def require(path)
-    required = begin
-      require_without_truffle_patching path
-    rescue LoadError
-      if Truffle::Patching::TRUFFLE_PATCHES[path]
-        true # Pretend the file required so we attempt patching with something that should avoid the LoadError (e.g., replacing code using C extensions).
-      else
-        raise
-      end
-    end
+  def gem_original_require(path)
+    extname        = File.extname path
+    path_no_suffix = path[0..(-1-extname.size)]
+    relative_path  = if path_no_suffix.start_with?('/')
```

Comment:
There is no such method, it's on Pathname. I'd rather not require more of stdlib here.